presto-main/src/main/java/com/facebook/presto/operator/TableScanOperator.java

```diff
@@ -197,6 +198,9 @@ public void finish()
                 throw Throwables.propagate(e);
             }
         }
+
+        operatorContext.getSystemMemoryTracker().freeMemory(recordedSystemMemoryUsage);
```

Comment:
Don't assume the source freed all its memory. Get its memory usage and only free the delta, since it's not yet eligible for garbage collection (see my other comment about `isFinished`)